File: phdf5-1/F90/phdf5_ex1.f90

! ###################################################################
! #                       HDF5 hands on                             #
! #  Created : April 2011                                           # 
! #                                                                 #
! #  Author:                                                        #
! #     Matthieu Haefele                                            #
! #     matthieu.haefele@ipp.mpg.de                                 #
! #     High Level Support Team (HLST)                              #
! #     Max-Planck Institut fuer Plasmaphysik                       #
! #                                                                 #
! ###################################################################

PROGRAM PHDF5_EX
     USE HDF5
     use mpi

     IMPLICIT NONE

     CHARACTER(LEN=10), PARAMETER :: filename = "example.h5" ! File name
     CHARACTER(LEN=8), PARAMETER :: dsetname = "IntArray"     ! Dataset name
     CHARACTER(LEN=15) :: obj_name 
     INTEGER, PARAMETER     ::   rank = 2                        
     INTEGER, PARAMETER     ::   NX = 16                        
     INTEGER, PARAMETER     ::   NY = 32                        

     INTEGER(HID_T) :: file, dataset, dataspace
     INTEGER(HSIZE_T), DIMENSION(2) :: dims
     INTEGER     ::   status
     INTEGER, DIMENSION(:,:), allocatable :: dat
     INTEGER     :: i,j,var_nx, var_ny, size_x, size_y
     integer :: mpi_id, size, ierr
    
     CALL MPI_INIT( ierr )
     CALL MPI_COMM_RANK( MPI_COMM_WORLD, mpi_id, ierr )
     CALL MPI_COMM_SIZE( MPI_COMM_WORLD, size, ierr )

     if(size .ne. 1 .and. size .ne. 4 .and. size .ne. 16) then
       if(mpi_id==0) then
         print*, "Error: Authorized number of process is 1, 4 or 16, exiting..."
       end if 
       call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
     end if

     var_nx = NX
     var_ny = NY
     call init(var_nx, var_ny, mpi_id, size, size_x, size_y, dat) 
        
     !
     ! Initialize FORTRAN predefined datatypes.
     !
     CALL h5open_f(status)

     do i=0, size-1
       if(i==mpi_id) then
         write(obj_name,'("example",I0.2,".h5")'),i
         !HDF5 file creation
         CALL h5fcreate_f(obj_name, H5F_ACC_TRUNC_F, file, status)

         !Dataspace creation
         dims(1) = size_x 
         dims(2) = size_y
         CALL h5screate_simple_f(rank, dims, dataspace, status)

         !Dataset creation 
         CALL h5dcreate_f(file, dsetname, H5T_NATIVE_INTEGER, dataspace, &
         dataset, status)

         !Actual data IO
         CALL h5dwrite_f(dataset, H5T_NATIVE_INTEGER, dat, dims, status)

         !Closing all opened HDF5 objects
         CALL h5dclose_f(dataset, status)
         CALL h5sclose_f(dataspace, status)
         CALL h5fclose_f(file, status)
       end if
       call MPI_Barrier(MPI_COMM_WORLD, ierr)
     end do

     CALL h5close_f(status)

     CALL MPI_FINALIZE(ierr)


     contains
subroutine init(NX, NY, mpi_id, size, size_x, size_y, dat)
  integer, intent(in) :: NX, NY, mpi_id, size
  integer, intent(out) :: size_x, size_y
  integer :: i,j
  integer, dimension(:,:), allocatable :: dat
  integer :: nb_proc_per_dim, offset_x, offset_y 
  nb_proc_per_dim = int(sqrt(real(size)))
  size_x = NX/nb_proc_per_dim
  size_y = NY/nb_proc_per_dim
  offset_x = mod(mpi_id,nb_proc_per_dim) * size_x
  offset_y = int(mpi_id/nb_proc_per_dim) * size_y

  allocate(dat(size_x,size_y))

  do j=1, size_y
    do i=1, size_x
      dat(i,j) = i+offset_x-1 + (j+offset_y-1)*NX;
    end do
  end do

end subroutine init

END PROGRAM PHDF5_EX
     

File: phdf5-1/C/solution/phdf5_ex1.c
/* ###################################################################
   #                       HDF5 hands on                             #
   #  Created : March 2015                                           #
   #                                                                 #
   #  Author:                                                        #
   #     Matthieu Haefele                                            #
   #     matthieu.haefele@maisondelasimulation.fr                    #
   #     CNRS/Maison de la Simulation                                #
   #                                                                 #
   ################################################################### */

#include <mpi.h>
#include <hdf5.h>
#include <math.h>
#include <stdlib.h>
#define NX     16
#define NY     32
#define RANK   2

int* init(int mpi_id, int size, int* sx, int* sy)
{
  int i,j;
  int* data;
  int nb_proc_per_dim = (int) sqrt(size);
  int size_x = NX/nb_proc_per_dim;
  int size_y = NY/nb_proc_per_dim;
  int offset_x = (mpi_id%nb_proc_per_dim) * size_x;
  int offset_y = ((int) mpi_id/nb_proc_per_dim) * size_y;

  data = malloc(size_x*size_y*sizeof(int));

  for(j=0; j < size_y; j++)
    for(i=0; i < size_x; i++)
      data[i + j*size_x] = i+offset_x + (j+offset_y)*NX;

  *sx = size_x;
  *sy = size_y;
  return data;
}

int main (int argc, char** argv)
{
  hid_t       file, dataset, dataspace;
  hsize_t     dimsf[RANK];
  herr_t      status;
  int*        data;
  int mpi_id, size;
  int size_x, size_y, i;
  char name[128];
  
  MPI_Init (&argc, &argv);
  MPI_Comm_rank (MPI_COMM_WORLD, &mpi_id);
  MPI_Comm_size (MPI_COMM_WORLD, &size);
  //printf( "Hello from process %d of %d\n", mpi_id, size );

  if(size != 1 && size != 4 && size != 16)
  {
    if(mpi_id==0)
      printf("Error: Authorized number of process is 1, 4 or 16, exiting...\n");
    MPI_Abort(MPI_COMM_WORLD,1);
  }


  //Data initialization
  data = init(mpi_id, size, &size_x, &size_y);
  printf("id=%d size_x=%d size_y=%d\n", mpi_id, size_x, size_y);
  
  if(mpi_id == 0)
  {
    //HDF5 file creation
    file = H5Fcreate("example.h5", H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
    dimsf[0] = size_y;
    dimsf[1] = size_x;

    //Dataspace creation
    dataspace = H5Screate_simple(RANK, dimsf, NULL); 

    //Dataset creation 
    sprintf(name, "IntArray%d", mpi_id);
    dataset = H5Dcreate(file, name, H5T_NATIVE_INT, dataspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);

    //Actual data IO
    status = H5Dwrite(dataset, H5T_NATIVE_INT, H5S_ALL, H5S_ALL,H5P_DEFAULT, data);

    //Closing all opened HDF5 objects
    H5Sclose(dataspace);
    H5Dclose(dataset);
    H5Fclose(file);
  }
  MPI_Barrier(MPI_COMM_WORLD);

  for(i=1; i < size; ++i)
  {
    if(i==mpi_id)
    {
      //HDF5 file creation
      file = H5Fopen("example.h5", H5F_ACC_RDWR, H5P_DEFAULT);
      dimsf[0] = size_y;
      dimsf[1] = size_x;

      //Dataspace creation
      dataspace = H5Screate_simple(RANK, dimsf, NULL); 

      //Dataset creation 
      sprintf(name, "IntArray%d", mpi_id);
      dataset = H5Dcreate(file, name, H5T_NATIVE_INT, dataspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);

      //Actual data IO
      status = H5Dwrite(dataset, H5T_NATIVE_INT, H5S_ALL, H5S_ALL,H5P_DEFAULT, data);

      //Closing all opened HDF5 objects
      H5Sclose(dataspace);
      H5Dclose(dataset);
      H5Fclose(file);
    }
    MPI_Barrier(MPI_COMM_WORLD);
  }
  
  MPI_Finalize();

  return 0;
}     
